### PR TITLE
Apl 833 implement unified cache service on node

### PR DESCRIPTION
**Pull request by alzinchenko**
Implemented unified cache subsystem. (Without visualization of cache statistics);
increased the guava version up to 28.0;

### CODE REVIEW

File: apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/cache/CacheConfigurator.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    private int maxSize = -1;
+    private CacheBuilder cacheBuilder;
+    private Optional<CacheLoader<K, V>> optionalCacheLoader;
+
+    public CacheConfigurator(String name, long elementSize, int cachePriority) {
+        this(name, elementSize, cachePriority, null);
+    }
+
+    public CacheConfigurator(String name, long elementSize, int cachePriority, CacheLoader<K, V> cacheLoader) {
+        this.name = name;
```
Comment (yuriy-larin-firstbridge) [inline apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/cache/CacheConfigurator.java]: I would add NULL check for 'name'.

File: apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/cache/CacheConfigurator.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        this(name, elementSize, cachePriority, null);
+    }
+
+    public CacheConfigurator(String name, long elementSize, int cachePriority, CacheLoader<K, V> cacheLoader) {
+        Objects.requireNonNull(name, "Cache name is NULL.");
+        this.name = name;
+        this.elementSize = elementSize;
+        this.cachePriority = cachePriority;
+        this.cacheBuilder = CacheBuilder.newBuilder();
+        this.optionalCacheLoader = Optional.ofNullable(cacheLoader);
```
Comment (AndrewBoyarsky) [inline apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/cache/CacheConfigurator.java]: Looks like Optional here is not required. Just return Optional.ofNullable in getCacheLoader

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/cache/ShardDsCacheConfigurator.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import com.apollocurrency.aplwallet.apl.util.cache.CacheConfigurator;
+
+import java.util.concurrent.TimeUnit;
+
+public class ShardDsCacheConfigurator extends CacheConfigurator {
+
+    public static final String SHARD_DS_CACHE_NAME = "SHARD_DS_CACHE";
+
+    public ShardDsCacheConfigurator(int percentCapacity) {
+        super(SHARD_DS_CACHE_NAME, 8, percentCapacity);
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/cache/ShardDsCacheConfigurator.java]: 8 - is not an actual size of datasource

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/cache/PublicKeyCacheConfig.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+package com.apollocurrency.aplwallet.apl.core.cache;
+
+import com.apollocurrency.aplwallet.apl.util.cache.CacheConfigurator;
+
+public class PublicKeyCacheConfig extends CacheConfigurator {
+
+    public static final String PUBLIC_KEY_CACHE_NAME = "PUBLIC_KEY_CACHE";
+
+    public PublicKeyCacheConfig(int percentCapacity) {
+        super(PUBLIC_KEY_CACHE_NAME, 32, percentCapacity);
```
Comment (AndrewBoyarsky) [inline apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/cache/PublicKeyCacheConfig.java]: 32 - size of public key array, not a size of array object itself

File: apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/cache/CacheConfigurator.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+public class CacheConfigurator<K, V> implements CacheConfiguration<K, V> {
+    private String name;
+    private long elementSize;
+    private int cachePriority;
+    private int maxSize = -1;
+    private CacheBuilder cacheBuilder;
+    private Optional<CacheLoader<K, V>> optionalCacheLoader;
+
+    public CacheConfigurator(String name, long elementSize, int cachePriority) {
```
[the remark this answers is not in the document]
Comment (alzinchenko) [reply apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/cache/CacheConfigurator.java]: no, it's priority

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/cache/ShardDsCacheConfigurator.java (hunk 3 above)
Comment (alzinchenko) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/cache/ShardDsCacheConfigurator.java]: this config is just an example

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/cache/PublicKeyCacheConfig.java (hunk 4 above)
Comment (alzinchenko) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/cache/PublicKeyCacheConfig.java]: yes, I'll resolve it in next PR.

File: apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/cache/InMemoryCacheManager.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+public class InMemoryCacheManager {
+
+    private static final long MIN_MEMORY_SIZE_FOR_CACHES = 16*1024*1024;//in bytes
+
+    private ConcurrentHashMap<String, Cache> inMemoryCaches;
+
+    @Inject
+    public InMemoryCacheManager(InMemoryCacheConfigurator configurator) {
+        Objects.requireNonNull(configurator, "Configurator is NULL.");
+        validateConfigurations(configurator);
```
[the remark this answers is not in the document]
Comment (alzinchenko) [reply apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/cache/InMemoryCacheManager.java]: @Postconstruct maybe not available in some environment.

File: apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/cache/InMemoryCacheManager.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        allocateAllCaches(configurator);
+    }
+
+    @Produces @CacheProducer
+    public <K, V> Cache<K, V> createCache(InjectionPoint injectionPoint){
```
[the remark this answers is not in the document]
Comment (alzinchenko) [reply apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/cache/InMemoryCacheManager.java]: I'll rename in next PR

File: apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/cache/CacheConfigurator.java (hunk 2 above)
Comment (alzinchenko) [reply apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/cache/CacheConfigurator.java]: I'll return Optional in that method.

File: apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/cache/InMemoryCacheManager.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    public <K, V> Cache<K, V> createCache(String cacheName){
+        Cache<K,V> cache = inMemoryCaches.get(cacheName);
+        if(cache == null){
+            log.warn("Cache [{}] not found.", cacheName);
+        }
+        return cache;
+    }
+
+    private void validateConfigurations(InMemoryCacheConfigurator cfg){
+        Preconditions.checkState(cfg.getAvailableMemory()>=MIN_MEMORY_SIZE_FOR_CACHES
```
[the remark this answers is not in the document]
Comment (alzinchenko) [reply apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/cache/InMemoryCacheManager.java]: Cache can't be optional, cause performance of blockchain is unacceptable without it.

File: apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/cache/InMemoryCacheManager.java [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+                cache = builder.build();
+            }
+            inMemoryCaches.put(config.getCacheName(), cache);
+            log.debug("Allocated cache={}", config);
+        });
+    }
+
+    private CacheBuilder configureCache(CacheConfiguration config, long availableMemory, int sumPriority){
+        //       key#hashCode + value#reference
+        int extra = 4         +      8;
```
[the remark this answers is not in the document]
Comment (alzinchenko) [reply apl-utils/src/main/java/com/apollocurrency/aplwallet/apl/util/cache/InMemoryCacheManager.java]: I'll add new MemoryUsageCalculator to help calculate memory usage.

File: apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/cache/PublicKeyCacheConfig.java (hunk 4 above)
Comment (alzinchenko) [reply apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/cache/PublicKeyCacheConfig.java]: I'll add new MemoryUsageCalculator to help calculate memory usage.

**Pull request merged**